### PR TITLE
Persist unknown settings.yaml entries across load/save.

### LINKED ISSUE #13142: Unknown settings should be stored and resaved
Our settings parser ignores any yaml entries that do not map to defined options, and so these are lost when the settings are reserialized.  This is a major problem for forward and backward compatibility now that we are aiming to support parallel installs.

### PR DESCRIPTION
**Pull request by pchote**
Closes #13142.

### CODE REVIEW

File: OpenRA.Game/Settings.cs [hunk 1]
```diff
@@ -392,11 +399,20 @@ public Settings(string file, Arguments args)
...
+			{
+				var sectionYaml = yamlCache.FirstOrDefault(x => x.Key == kv.Key);
+				if (sectionYaml == null)
+				{
+					sectionYaml = new MiniYamlNode(kv.Key, new MiniYaml(""));
+					yamlCache.Add(sectionYaml);
+				}
+
+				var currentSection = FieldSaver.SaveDifferences(kv.Value, Activator.CreateInstance(kv.Value.GetType()));
+				sectionYaml.Value.Nodes = MiniYaml.Merge(new[] { sectionYaml.Value.Nodes, currentSection.Nodes });
```
Comment (pchote) [inline OpenRA.Game/Settings.cs]: If `sectionYaml` contains a custom value and the player changes something back to its default then `currentSection` won't have an override, and the new value won't overwrite the old.

**Comment by pchote** [comment]
Fixed.

**Comment by reaperrr** [comment]
The code is out of my league so I'd like either @RoosterDragon or @atlimit8 to approve this before it's merged, but I can at least confirm that it works as advertised. 👍

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
When switching back from this branch to another one, the game wrote a duplicate `Debug: True` key, which would then lead to a crash.

Not sure if that's a side-effect that can't be avoided or actually something that can/should be fixed.

**Comment by pchote** [comment]
I'm not sure how you'd trigger writing a duplicate entry.  Can you provide a repro case?

The `DictFromFile` will indeed throw on duplicate entries, which is the main reason why I removed it instead of leaving it unused for the future.

File: OpenRA.Game/Settings.cs [hunk 2]
```diff
@@ -392,11 +399,39 @@ public Settings(string file, Arguments args)
...
+					if (serialized == defaultSerialized)
+						sectionYaml.Value.Nodes.RemoveAll(n => n.Key == fli.YamlName);
+					else
+					{
+						// Update or add the custom value
+						var fieldYaml = sectionYaml.Value.Nodes.FirstOrDefault(n => n.Key == fli.YamlName);
+						if (fieldYaml != null)
+							fieldYaml.Value.Value = serialized;
+						else
+							sectionYaml.Value.Nodes.Add(new MiniYamlNode(kv.Key, new MiniYaml(serialized)));
```
Comment (rob-v) [inline OpenRA.Game/Settings.cs]: It needs to be: new MiniYamlNode(_fli.YamlName,_ ..., not kv.Key.

**Comment by pchote** [comment]
Updated.  The issue pointed out by @rob-v was the cause of @reaperrr's crash, which is now fixed.

**Review by rob-v — APPROVED** [review]

File: OpenRA.Game/Settings.cs [hunk 3]
```diff
@@ -391,11 +398,39 @@ public Settings(string file, Arguments args)
...
+				var fields = FieldLoader.GetTypeLoadInfo(kv.Value.GetType());
+				foreach (var fli in fields)
+				{
+					var serialized = FieldSaver.FormatValue(kv.Value, fli.Field);
+					var defaultSerialized = FieldSaver.FormatValue(defaultValues, fli.Field);
+
+					// Fields with their default value are not saved in the settings yaml
+					// Make sure that we erase any previously defined custom values
+					if (serialized == defaultSerialized)
+						sectionYaml.Value.Nodes.RemoveAll(n => n.Key == fli.YamlName);
```
Comment (atlimit8) [inline OpenRA.Game/Settings.cs]: I'm just noting that if a user changes a setting to a value that happens to be the current default, they might be surprised if it changes, which would only occur if the default value changes between versions.

**Review by atlimit8 — APPROVED** [review]

**Comment by atlimit8** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/684f3b03983929a2f516cd75fc12efc9ae227196...1a21847ac9b116d83651ec8559169201c8aa7)

**Pull request merged**